lnwallet/rpcwallet/rpcwallet.go

```diff
@@ -401,48 +219,146 @@ func (r *RPCKeyRing) FinalizePsbt(packet *psbt.Packet, accountName string) error
 	return nil
 }
 
-// DeriveNextKey attempts to derive the *next* key within the key family
-// (account in BIP43) specified. This method should return the next external
-// child within this branch.
+// FinalizePsbt expects a partial transaction with all inputs and outputs fully
+// declared and tries to sign all inputs that belong to the specified account.
+// Lnd must be the last signer of the transaction. That means, if there are any
+// unsigned non-witness inputs or inputs without UTXO information attached or
+// inputs without witness data that do not belong to lnd's wallet, this method
+// will fail. If no error is returned, the PSBT is ready to be extracted and the
+// final TX within to be broadcast.
 //
-// NOTE: This method is part of the keychain.KeyRing interface.
-func (r *RPCKeyRing) DeriveNextKey(
-	keyFam keychain.KeyFamily) (keychain.KeyDescriptor, error) {
+// NOTE: This method does NOT publish the transaction after it's been
+// finalized successfully.
+//
+// NOTE: This is a part of the WalletController interface.
+//
+// NOTE: We need to overwrite this method because we need to redirect the call
+// to ComputeInputScript to the RPC key ring's implementation. If we forward
+// the call to the default WalletController implementation, we get an error
+// since that wallet is watch-only. If we forward the call to the remote signer,
+// we get an error because the signer doesn't know the UTXO information required
+// in ComputeInputScript.
+//
+// TODO(guggero): Refactor btcwallet to accept ComputeInputScript as a function
+// parameter in FinalizePsbt so we can get rid of this code duplication.
+func (r *RPCKeyRing) FinalizePsbt(packet *psbt.Packet, _ string) error {
+	// Let's check that this is actually something we can and want to sign.
+	// We need at least one input and one output.
+	err := psbt.VerifyInputOutputLen(packet, true, true)
+	if err != nil {
+		return err
+	}
+
+	// Go through each input that doesn't have final witness data attached
+	// to it already and try to sign it. We do expect that we're the last
+	// ones to sign. If there is any input without witness data that we
+	// cannot sign because it's not our UTXO, this will be a hard failure.
+	tx := packet.UnsignedTx
+	sigHashes := txscript.NewTxSigHashes(tx)
+	for idx, txIn := range tx.TxIn {
+		in := packet.Inputs[idx]
+
+		// We can only sign if we have UTXO information available. We
+		// can just continue here as a later step will fail with a more
+		// precise error message.
+		if in.WitnessUtxo == nil && in.NonWitnessUtxo == nil {
```

Comment:
Can we have both as non-nil?

Reply:
Yes, by default we always set both in `lnd`. Some hardware wallets even require both to be set for a witness input (there was a CVE a while ago that made everyone patch the signing software to require the full non-witness UTXO information to be present).